Autogenerate Java docs (II)

dgarcia360 opened the pull request:
Summarizes all commits from https://github.com/scylladb/java-driver/pull/29 in just one commit

dgarcia360 committed:
Docs automation

Update README-dev.rst

Prefer warnings

Updated CHANGELOG

Import README from original file

Rollback spaces

Fix spaces

Fix spaces

Added 404

Removed  pipx

haaawk commented:
This looks mostly good to me but it's also very invasive and will probably generate lots of merge conflicts once we sync with the upstream. 

1. Are all those changes necessary? @tzach 
2. What's the procedure to use those scripts?

dgarcia360 commented:
@haaawk How to build the docs is explained in this file https://github.com/scylladb/java-driver/pull/38/files#diff-3160134a4c3b6772a52ddbc542e75d2f

tzach commented on upgrade_guide/migrating_from_astyanax/queries_and_results/README.md:
````diff
@@ -1,3 +1,7 @@
+```eval_rst
+:orphan:
+```
````
Are this addition a must?
It will make sync with upstream hard
If no other way, lets disable the warning it generate

dgarcia360 replied:
It's not a must, but Sphinx throws the warning ``WARNING: document isn't included in any toctree``. This does not stop the documentation from the building if deleted.

tzach commented:
@haaawk The goal is making the MD file update minimal to make sync with upstream easy
We are almost there

haaawk commented:
> @haaawk How to build the docs is explained in this file https://github.com/scylladb/java-driver/pull/38/files#diff-3160134a4c3b6772a52ddbc542e75d2f

Ok. I asked the wrong question. What do I have to do to cherry-pick this commit to a new branch if I need to (and I know for sure I will have to do this many times)?

haaawk commented on faq/osgi/README.md:
````diff
@@ -1,7 +1,3 @@
-```eval_rst
````
Please squash this commit with the first commit in the PR. Otherwise the first commit will create conflicts anyway.

dgarcia360 commented:
@haaawk I'd like to understand the process you follow to pull the latest changes from upstream while keeping all the modifications you applied before. Could you share with me the steps you follow in the terminal?  With this info, I could write some instructions showing how we can apply the docs generation to the next releases. Thanks!

haaawk commented on changelog/README.md:
```diff
@@ -46,7 +46,7 @@
 
 - [bug] JAVA-2174: Metadata.needsQuote should accept empty strings.
-
+- [bug] JAVA-2193: Fix flaky tests in WarningsTest.
```
Where does this come from?

dgarcia360 replied:
It was deleted in a previous commit and re-added in this one, there are no changes.

haaawk replied:
Please squash it into the previous commit that removes it.

haaawk commented:
> @haaawk I'd like to understand the process you follow to pull the latest changes from upstream while keeping all the modifications you applied before. Could you share with me the steps you follow in the CLI? With this info, I could write some instructions showing how we can apply the docs generation to the next releases. Thanks!

Thanks @dgarcia360. So far I was only doing this with the Java code and the process is as follows:
1. I have an old branch which is based on some upstream branch. Let's call the branch 3.7.1-scylla and the upstream branch 3.7.1. This branch has some Scylla specific modifications.
2. When there's a new release of the upstream. For example 3.8.0. We fork it, create new branch called 3.8.0-scylla and cherry-pick all the Scylla specific modifications to it.

I use `git cherry-pick` to cherry pick the changes. Most of the time they apply cleanly but sometimes there are merge conflicts which I resolve manually. I know how to do that because it's just code.
Now the question is what can I do for documentation in case of conflict.
One option is resolving every conflict manually. I can do it for docs but I'm not sure about the scripts.
Is there any other option?

dgarcia360 commented:
Thanks @haaawk. The problem with the current process applies to code and docs in the same way. If you have to edit the custom code, you would have to cherry-pick multiple commits in order to apply the changes to newer versions, which will become complex to manage if many changes are added.

I propose to merge all changes in one integration branch (e.g. ``latest``). As far as I know, other drivers like https://github.com/scylladb/python-driver/ already follow this approach.

- If there is a custom modification, you can add it to the integration branch without having to cherry-pick this commits in the future.
- If there is a new upstream release, you can move all the changes from their version branch into your ``latest`` with ``git merge upstream/<version-name>``. This will require resolving all the merge conflicts that might arise due to the modifications.

Once the branch is in a status to be released, you can either create a new branch from the integration branch and tag it with the version number or create a GitHub release.

Please let me know if this new process could work in this repository. If not, the other alternative would be to cherry-pick this PR, which has already been squashed in commit, and pick other future docs modifications as now is being done with code changes.

dgarcia360 commented:
I had a conversation today with @fruch to know more about how he is managing the python-driver releases. I can confirm that the python driver merges all the changes into an integration branch (master) to advance on the documentation efforts and other custom code added to the plugin advance if necessary.

@haaawk Here is [the script](https://github.com/scylladb/python-driver/blob/master/scripts/merge_next_tag_from_upstream.sh) used by the python-driver to merge the next tag from the upstream. Could you please review it and evaluate if it would be possible to use the same approach for the java-driver?

I'm totally aware that this would require some extra efforts to merge new releases that might not have been originally scoped in the short term. However, forks are normally created to extend or adapt the functionality, fix some bugs, or add additional value. Looking into the future, and assuming that new changes would be added as the time passes (code and docs), it would cause even fewer conflicts integrating the latest commits into an integration branch instead of cherry-picking 100s of commits.

We could also use the script created by @fruch to set an internal standard to work with projects depending on upstream repos. Consistency fastens the onboarding of new contributors to the driver's projects, and the integration can happen regularly since downloading the latest changes can be partially automated.

Other solutions have been discussed, like treating docs generation exclusively in a separate orphan branch or abstracting the docs generation in a separate repository. The first would require to deal with merge conflicts every time we want to update the documentation branch, deferring but not solving the problem to treat with merges. The second would work well to generate docs, but editing any documentation file (e.g. Rellabellign DatasTax -> ScyllaDB) would not be possible without editing this repository.

haaawk commented:
@fruch how is this script working for you? I can see that the last changes pulled from upstream in python driver were on 26th Feb 2020. Documentation changes were done only after that. Have you tried running the script after documentation changes? How many conflicts did it generate?

I'm a bit sceptical to the integration branch idea. We have a few changes in the driver we maintain while the upstream is actively developed by a whole team of people. The tradeoff is definitely in favor of cherry-picking a few commits than resolving conflicts when merging 100s of new commits coming from the upstream. 
The integration branch may work for python because it has way fewer active contributions:
python
![image](https://user-images.githubusercontent.com/1311149/95770428-ac0dca00-0cb9-11eb-8021-63ecc5ec76e0.png)
java
![image](https://user-images.githubusercontent.com/1311149/95770492-c34cb780-0cb9-11eb-9726-32ab88f543fc.png)
Moreover, Java driver is released multiple times a month and have two flavors 3.x and 4.x
![image](https://user-images.githubusercontent.com/1311149/95771025-a49af080-0cba-11eb-812f-c6db20d7335f.png)
While Python driver is released much less frequently
![image](https://user-images.githubusercontent.com/1311149/95772019-5d156400-0cbc-11eb-88bc-2bba0f8cd9ab.png)



Thus, I'd rather have docs generation on a separate branch or in a separate repository.
Another problem is that users sometimes need Scylla Java driver based on a very specific version of upstream. The integration branch does not allow it.

fruch commented:
@haaawk I'm not sure I'm following, you are reviewing each patch an cherry pick them one by one into a branch ? And anyhow you need to solve call conflict if there are any...

We decided to keep releasing version based on the upstream versions, and we only have patch version in between.

As for custom versions, we are not yet been asked for such things, so far people were taking only one the versions we released.

haaawk commented:
@fruch When I create new release based on some upstream branch X, I first create a new branch based on X and then I cherry-pick our Scylla-specific changes on top of it.

fruch commented:
> @fruch When I create new release based on some upstream branch X, I first create a new branch based on X and then I cherry-pick our Scylla-specific changes on top of it.

so have 1 or 2 of those for docs, and it's basically the same.

I think docs should be first class citizen in a code repo.

haaawk commented on docs/source/installation.md:
```diff
@@ -0,0 +1,76 @@
+If you are upgrading from a previous version of the driver, be sure to have a look at
+the [upgrade guide](upgrade_guide/).
+
+If you are upgrading to _DataStax Enterprise_, use the [DataStax Enterprise Java driver][dse-driver] for more
+features and better compatibility.
```
This shouldn't be here.

dgarcia360 replied:
The PR only adds documentation generation automation. The file ``installation.md`` was created copying the contents from the README to match the python-driver documentation structure. Once this is merged, please coordinate with @lauranovich to relabel the whole docs to ScyllaDB.

dgarcia360 replied:
@haaawk The README file is now imported from the original file https://github.com/scylladb/java-driver/pull/38/files. The changes can be edited directly by editing https://github.com/scylladb/java-driver/blob/latest/README.md

haaawk commented on docs/source/installation.md:
```diff
@@ -0,0 +1,76 @@
+If using _DataStax Enterprise_, the [DataStax Enterprise Java driver][dse-driver] provides 
+more features and better compatibility.
+
+__Disclaimer__: Some _DataStax/DataStax Enterprise_ products might partially work on 
+big-endian systems, but _DataStax_ does not officially support these systems.
```
This shouldn't be here.

haaawk commented on docs/source/installation.md:
```diff
@@ -0,0 +1,76 @@
+
+The java driver supports Java JDK versions 6 and above.
+
+If using _DataStax Enterprise_, the [DataStax Enterprise Java driver][dse-driver] provides 
+more features and better compatibility.
```
This shouldn't be here

haaawk commented on docs/source/installation.md:
```diff
@@ -0,0 +1,76 @@
+
+### Compatibility
+
+The Java client driver 3.7.1 ([branch 3.x](https://github.com/datastax/java-driver/tree/3.x)) is compatible with Apache
+Cassandra 2.1, 2.2 and 3.0+ (see [this page](http://docs.datastax.com/en/developer/java-driver/latest/manual/native_protocol/) for
+the most up-to-date compatibility information).
+
+UDT and tuple support is available only when using Apache Cassandra 2.1 or higher (see [CQL improvements in Cassandra 2.1](http://www.datastax.com/dev/blog/cql-in-2-1)).
+
+Other features are available only when using Apache Cassandra 2.0 or higher (e.g. result set paging,
+[BatchStatement](https://github.com/datastax/java-driver/blob/3.x/driver-core/src/main/java/com/datastax/driver/core/BatchStatement.java),
+[lightweight transactions](http://www.datastax.com/documentation/cql/3.1/cql/cql_using/use_ltweight_transaction_t.html) 
+-- see [What's new in Cassandra 2.0](http://www.datastax.com/documentation/cassandra/2.0/cassandra/features/features_key_c.html)). 
+Trying to use these with a cluster running Cassandra 1.2 will result in 
+an [UnsupportedFeatureException](https://github.com/datastax/java-driver/blob/3.x/driver-core/src/main/java/com/datastax/driver/core/exceptions/UnsupportedFeatureException.java) being thrown.
```
All those links are to the wrong repository.

haaawk commented on docs/source/index.rst:
```diff
@@ -0,0 +1,100 @@
+
+**Useful links:**
+
+-  JIRA (bug tracking): https://datastax-oss.atlassian.net/browse/JAVA
+-  MAILING LIST:
+   https://groups.google.com/a/lists.datastax.com/forum/#!forum/java-driver-user
+-  DATASTAX ACADEMY SLACK: #datastax-drivers on
+   https://academy.datastax.com/slack
+-  TWITTER: `@dsJavaDriver`_ tweets Java driver releases and important
+   announcements (low frequency). `@DataStaxEng`_ has more news
+   including other drivers, Cassandra, and DSE.
+-  DOCS: the `manual`_ has quick start material and technical details
+   about the driver and its features.
+-  API: http://www.datastax.com/drivers/java/3.7
+-  GITHUB REPOSITORY: https://github.com/datastax/java-driver
```
All those links are wrong.

haaawk commented on docs/source/index.rst:
```diff
@@ -0,0 +1,100 @@
+Datastax Java Driver for Apache Cassandra®
```
s/Datastax/Scylla

haaawk commented:
> > @fruch When I create new release based on some upstream branch X, I first create a new branch based on X and then I cherry-pick our Scylla-specific changes on top of it.
> 
> so have 1 or 2 of those for docs, and it's basically the same.

It's not. I'm capable of resolving conflicts in code but I know nothing about this docs generation mechanism.

> 
> I think docs should be first class citizen in a code repo.

Sure but automation like this needs to be maintained by someone who actually understands what it does. Who's it gonna be? @lauranovich ?

haaawk commented:
I just tried to cherry-pick this PR on top of 4.x:
```
$ git status
On branch test
Your branch is up to date with 'origin/4.x'.

You are currently cherry-picking commit e0e51e185.
  (fix conflicts and run "git cherry-pick --continue")
  (use "git cherry-pick --skip" to skip this patch)
  (use "git cherry-pick --abort" to cancel the cherry-pick operation)

Changes to be committed:
	new file:   .github/workflows/pages.yml
	new file:   README-dev.rst
	new file:   docs/Makefile
	new file:   docs/_utils/deploy.sh
	new file:   docs/_utils/doxygen.sh
	new file:   docs/_utils/multiversion.sh
	new file:   docs/_utils/redirect.sh
	new file:   docs/_utils/redirections.yaml
	new file:   docs/_utils/setup.sh
	new file:   docs/poetry.lock
	new file:   docs/pyproject.toml
	new file:   docs/source/api.rst
	new file:   docs/source/changelog
	new file:   docs/source/conf.py
	new file:   docs/source/faq
	new file:   docs/source/index.rst
	new file:   docs/source/installation.md
	new file:   docs/source/manual
	new file:   docs/source/upgrade_guide

Unmerged paths:
  (use "git add/rm <file>..." as appropriate to mark resolution)
	both modified:   .gitignore
	both modified:   changelog/README.md
	deleted by them: docs.yaml
	deleted by us:   doxyfile
	both modified:   manual/README.md
	deleted by us:   manual/object_mapper/README.md
	deleted by us:   manual/statements/README.md
	both modified:   upgrade_guide/README.md
	deleted by us:   upgrade_guide/migrating_from_astyanax/README.md
```

What am I suppose to do with `deleted by us:   doxyfile`?

fruch commented:
> > > @fruch When I create new release based on some upstream branch X, I first create a new branch based on X and then I cherry-pick our Scylla-specific changes on top of it.
> > 
> > 
> > so have 1 or 2 of those for docs, and it's basically the same.
> 
> It's not. I'm capable of resolving conflicts in code but I know nothing about this docs generation mechanism.
> 
> > I think docs should be first class citizen in a code repo.
> 
> Sure but automation like this needs to be maintained by someone who actually understands what it does. Who's it gonna be? @lauranovich ?

It's called sphinx and it's the same one used across all our docs (from scylla core ones, to python-driver)
@dgarcia360 expertise was called for this purpose

So the answer who maintains it in all those other places would help, plus I can give a hand with it if needed, I have some mileage with it.

dgarcia360 commented:
The latest commit imports the README contents from the original file https://github.com/scylladb/java-driver/pull/38/files to have fewer changes.

> Thus, I'd rather have docs generation on a separate branch

Ok, let's start by integrating the changes into a separate branch. At least we will be able to serve the docs for the current version while keeping the door open to move to the integration branch approach in future iterations.  Once there are new upstream releases, we will be able to test in this branch how difficult is to merge the changes in the documentation branch and evolve the solution from there.

@haawk Could you please create a new branch named ``docs`` aligned with the latest version? I'll move the base and edit the code to publish the docs from this new branch.

haaawk commented:
Sure. Will do.

dgarcia360 commented:
Thanks!

lauranovich commented:
@haaawk let me know when you want me to step in - @dgarcia360 what else remains?

haaawk commented:
The branch is here https://github.com/scylladb/java-driver/tree/docs

haaawk commented:
Should we merge this PR against docs branch instead of latest?

haaawk commented:
I talked to @lauranovich and we came to the conclusion that we can try to accept this PR on the release branch. There are 2 things I need to ask you @dgarcia360 though:
1. Please make sure that you don't override changes you do in one commit in the following commit. This will make my life easier when I will be cherry-picking your commits to the new branches
2. Please prepare one version of this PR for 3.10.2-scylla and another for 4.9.0-scylla. The two lines (3.x and 4.x) differ significantly but inside a single line, I should be able to manage conflicts.

Thanks,
Piotr

dgarcia360 commented:
> I talked to @lauranovich and we came to the conclusion that we can try to accept this PR on the release branch

Nice! I rolled back the commit and moved the base to the ``latest`` branch.

> 1. Please make sure that you don't override changes you do in one commit in the following commit. This will make my life easier when I will be cherry-picking your commits to the new branches

Configuration-wise, I don't expect to add more changes other than minor commits to update the [theme's version](https://github.com/scylladb/java-driver/pull/38/files#diff-7a4371128da345a5bf5433ec7cd2fbb15e0e7ba303decbcc912a3e7617f0880dR12) and to list new supported documentation versions in the file [conf.py](https://github.com/scylladb/java-driver/pull/38/files#diff-008dcb3426febd767787b1521f1fe33086313b927ea37eaab86df5fa88a51698R182) from time to time.  

Based on other projects we have added docs, I don't think that the documentation contents will be kept static. At least some links pointing to upstream docs and texts will need edits by you and @lauranovich in future PRs.

> 2.Please prepare one version of this PR for 3.10.2-scylla and another for 4.9.0-scylla. The two lines (3.x and 4.x) differ significantly but inside a single line, I should be able to manage conflicts.

Ok, once the PR is merged I'll submit a couple of PRs to these specific branches by cherry-picking this commit. Let me know if you want to try yourself instead and do not hesitate to share with me any issues that may arise.
